Replace CAS Jepsen test with a Jepsen test of the timelock server

harrybiddle opened the pull request:
This PR removes the old Jepsen test, replacing it with one for the timelock server. The intention is that we will be able to re-visit the old test, and put it back into place.

Things this PR includes:

- Mount the `atlasdb-jepsen-test` directory, rather than `ADD`ing it.
- Compile `atlasdb-jepsen-test` into a fat jar so that it can be included in the JVM that runs the Jepsen test
- Add some Java code to create timestamp clients and wait for them to be ready, so that we don't have to write this in Clojure
- Use a newer Jepsen version (0.1.3)
- Move the awaitility version into `version.props`

Future work for the timelock Jepsen test that this PR does not include:

- Having it run at a reasonable speed, appropriate for our CI
- Enabling it on our CircleCI build
- Breaking the timelock server (e.g. add some noise to the timestamps) and ensuring that these tests fail appropriately

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/1305)
<!-- Reviewable:end -->

SerialVelocity commented on atlasdb-jepsen-tests/.gitignore:
```diff
@@ -1,2 +1,7 @@
+/.lein-failures
+/.nrepl-port
+/.lein-repl-history
+/target/
```
Can you run `M-x sort-lines` over this? (If you're an emacs user)

SerialVelocity commented on atlasdb-jepsen-tests/Dockerfile:
```diff
@@ -1,5 +1,3 @@
 FROM tjake/jepsen:latest
 
-ADD . /jepsen/atlasdb
```
If you're removing this, you can just override the command directly in the docker-compose file and not have this file at all.

SerialVelocity commented on atlasdb-jepsen-tests/build.gradle:
```diff
@@ -37,19 +41,30 @@ dependencies {
+    }
+}
+
+task copyFatJar(type: Copy, dependsOn: 'shadowJar') {
```
`copyShadowJar`?

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
+        (c/exec :mkdir "/atlasdb-timelock-server")
+        (c/exec :tar :xf "/atlasdb-timelock-server.tgz" "-C" "/atlasdb-timelock-server" "--strip-components" "1")
+        (c/upload "resources/atlasdb/timelock.yml" "/atlasdb-timelock-server/var/conf")
+        (c/exec :sed :-i (format "'s/HOSTNAME/%s/'" (name node)) "/atlasdb-timelock-server/var/conf/timelock.yml")
```
Can you make it `${HOSTNAME}` or `{{HOSTNAME}}` to make it clear it is going to be replaced?

harrybiddle replied:
Good idea: I've gone for `<HOSTNAME>`, since that played the nicest with the `c/exec :sed` line, and it can't be confused with it being an environment variable.

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
+      (c/su
+        (debian/install-jdk8!)
+        (c/upload "resources/atlasdb/atlasdb-timelock-server.tgz" "/")
+        (info node "Uploading and unpacking timelock server")
```
This should probably be above the upload

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
-        (teardown-cassandra node)))
+    (setup! [_ _ node]
+      (c/su
+        (debian/install-jdk8!)
```
Can we add an info for installing the jdk8 since it takes so long?

harrybiddle replied:
Jepsen already adds this

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
+        (c/upload "resources/atlasdb/timelock.yml" "/atlasdb-timelock-server/var/conf")
+        (c/exec :sed :-i (format "'s/HOSTNAME/%s/'" (name node)) "/atlasdb-timelock-server/var/conf/timelock.yml")
+        (info node "Starting timelock server")
+        (c/exec :env "JAVA_HOME=/usr/lib/jvm/java-8-oracle/" "/atlasdb-timelock-server/service/bin/init.sh" "start")
```
Doesn't installing jdk8 auto set JAVA_HOME properly?

harrybiddle replied:
Hmm, looks like it does, but `c/exec` isn't a login shell. I've changed this to `source /etc/profile`.

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
-        (jepsen/with-os test#
-          (jepsen/with-db test#
-            ~@body))))))
+      ["/atlasdb-timelock-server/var/log/atlasdb-timelock-server-startup.log"])))
```
I think there's also a `atlasdb-timelock-server.log` once the server has started

SerialVelocity replied:
and a requests log

harrybiddle replied:
I can't find those. In my runs, all I see are these two:

```
atlasdb-timelock-server-startup.log  
gc-2016-11-30_17-12-21-pid5343.log.0.current
```

SerialVelocity commented on atlasdb-jepsen-tests/src/jepsen/atlasdb.clj:
```diff
@@ -1,178 +1,99 @@
+  "
+  [timestamp-client]
+  (reify client/Client
+
```
nit: new line

harrybiddle replied:
done

SerialVelocity commented on atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/TimestampClient.java:
```diff
@@ -0,0 +1,90 @@
+        return Lists.transform(hosts, host -> String.format("http://%s:%d/%s", host, PORT, NAMESPACE));
+    }
+
+    public static TimestampService createFromUris(List<String> endpointUris) {
```
It's weird that this is all the way down here. It should be with the other `create` at the top

harrybiddle replied:
Actually, it can be private! I've made this private, and deleted `randomizeHostsAndCreate` too

SerialVelocity commented on atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/TimestampClient.java:
```diff
@@ -0,0 +1,90 @@
+                .until(() -> clusterReturnsTimestamp(hosts));
+    }
+
+    private static boolean clusterReturnsTimestamp(List<String> hosts) {
```
This should probably take in a `TimestampService`. Creating one every time seems odd.

harrybiddle replied:
This is used in a polling function that is waiting until the timelock server is ready. The first time it is called, the timelock server is still starting up. My motivation was that it would be better to keep re-creating the service than to create one before the timelock server was ready and keep using it, in case it got into a bad state somehow.

Do you still think it's better just to create one?

SerialVelocity commented on versions.props:
```diff
@@ -8,6 +8,7 @@ com.google.dagger:dagger = 2.0.2
 com.google.dagger:dagger-compiler = 2.0.2
 com.google.guava:* = 18.0
 com.googlecode.json-simple:json-simple = 1.1.1
+com.jayway.awaitility:awaitility = 1.6.5
```
:+1:

SerialVelocity commented on atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/TimestampClient.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public static TimestampService createFromUris(List<String> endpointUris) {
+        return AtlasDbHttpClients.createProxyWithFailover(Optional.<SSLSocketFactory>absent(), endpointUris,
```
You should really put all parameters on separate lines if you do it for one.

harrybiddle replied:
Done

harrybiddle commented:
Changes since last review:

- Removed docker-compose build step; just using tjake/jepsen:latest in docker-compose.yml
- Reduced the public API of TimestampClient: `createFromUris` is private; no more `randomizeHostsAndCreate`
- Now using `<HOSTNAME>` in the template timelock.yml
- Minor changes (sort .gitignore, copyFatJar -> copyShadowJar, newlines)
